docs/core/upgrade.md

```diff
@@ -0,0 +1,108 @@
+
+```golang
+type UpgradeHandler func(ctx sdk.Context, plan Plan, fromVM VersionMap) (VersionMap, error)
+```
+
+Inside these functions, you should perform any upgrade logic you wish to include in the provided `plan`. All upgrade handler functions should end with the following line of code:
+
+```golang
+  return app.mm.RunMigrations(ctx, cfg, fromVM)
+```
+
+## Running Migrations
+
+Migrations are run inside of an `UpgradeHandler` via `app.mm.RunMigrations(ctx, cfg, vm)`. As described above, `UpgradeHandler`s are functions which describe the functionality to occur during an upgrade. The `RunMigration` function will loop through the `VersionMap` argument, and run the migration scripts for any versions that are less than the new binary's app module versions. Once the migrations are finished, a new `VersionMap` will be returned to persist the upgraded module versions to state.
+
+```golang
+cfg := module.NewConfigurator(...)
+app.UpgradeKeeper.SetUpgradeHandler("my-plan", func(ctx sdk.Context, plan upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
+
+    // ...
+    // do upgrade logic
+    // ...
+
+    // RunMigrations returns the VersionMap
+    // with the updated module ConsensusVersions
+    return app.mm.RunMigrations(ctx, vm)
+})
+```
+
+To learn more about configuring migration scripts for your modules, refer to this [guide](../building-modules/upgrade.md).
+
+## Adding New Modules In Upgrades
+
+Entirely new modules can be introduced to the application during an upgrade. The SDK recognizes new modules during upgrades because their consensus version in the `fromVM` `VersionMap` is 0. In this case, `RunMigrations` will call the corresponding module's `InitGenesis` function to setup its initial state. This can be skipped if the module does not require any inital state. Otherwise, it is important to implement `InitGenesis` for new modules to successfully upgrade your application without error.
+
+In the scenario where your application does not need any inital state via `InitGenesis`, you must take extra steps to ensure `InitGenesis` is skipped to avoid errors. To do so, you simply need to update the value of the module version in the `VersionMap` in the `UpgradeHandler`. 
+
+```go
+// Foo is a new module being introduced
+// in this upgrade plan
+import foo "github.com/my/module/foo"
+
+app.UpgradeKeeper.SetUpgradeHandler("my-plan", func(ctx sdk.Context, plan upgradetypes.Plan, vm module.VersionMap)  (module.VersionMap, error) {
+    // We make sure to set foo's version to the latest ConsensusVersion in the VersionMap.
+    // This will skip running InitGenesis on Foo
+    vm["foo"] = foo.AppModule{}.ConsensusVersion()
+
+    return app.mm.RunMigrations(ctx, cfg, vm)
+})
+```
+
+Using a similar method, you can also run `InitGenesis` on your new module with a custom genesis state:
```

Comment:
so we can manually update the value of the module version in the `VersionMap` in `UpgradeHandler` (line 79) or we can run `InitGenesis` on your new module with a custom genesis state?
what are the use cases for doing one or the other? If there isn't a difference, I recommend that we give one method only. If there are use cases for or against, we should describe the benefits

Reply:
That's a good point.

For new modules, the default behavior is to run `InitGenesis` with the genesis state chosen by the _module developer_, which acts like a sane default. The module developer implements that in `AppModule.InitGenesis`.

However, there might be cases where the _app developer_ wants to overwrite that sane default. In this case, the app developer can update the module version in `VersionMap` and run `InitGenesis` with their own custom  genesis state.

I agree that the use cases for and against could be described here 👍

Reply:
Would this imply that, for an arbitrary upgrade, both the app dev and the module dev have written two genesis functions, which are now in competition for the upgrade? 

Couldn't the app developer just change the InitGenesis function in the module? Otherwise it feels a bit hacky and confusing imo. Thoughts @amaurym ?

Reply:
> Couldn't the app developer just change the InitGenesis function in the module?

I guess i'm thinking about the case where the app developer != the module developer, e.g. Regen importing x/authz in its next upgrade, but wants a custom genesis for x/authz. In the regen repo, we can't change x/authz's InitGenesis, so we must use the "setting the vm value" method to run a custom InitGenesis.

For modules where app developer == module developer, yeah it's nonsensical to write 2 genesis functions.

Reply:
ahhh okay thats what i was missing. i'll update the docs to reflect this

Reply:
[updated overwriting genesis seciton](https://github.com/cosmos/cosmos-sdk/pull/8967/commits/7883af3bee69ef1232c670271e4254ba2515c3f6)

@barriebyron @amaurym